pkg/checkpoint/process.go

```diff
@@ -9,138 +10,542 @@ import (
+		state := cs.selfCheckpoint.state.transition(now, apiState{
+			apiAvailable: apiAvailable,
+			apiParent:    apiParentPods[cs.selfCheckpoint.name] != nil,
+			localRunning: localRunningPods[cs.selfCheckpoint.name] != nil,
+			localParent:  localParentPods[cs.selfCheckpoint.name] != nil,
+		})
+
+		if state != cs.selfCheckpoint.state {
+			glog.Infof("Self-checkpoint %s transitioning from state %s -> state %s", cs.selfCheckpoint, cs.selfCheckpoint.state, state)
+			cs.selfCheckpoint.state = state
+		}
+
+		switch cs.selfCheckpoint.state.action() {
+		case none:
+			glog.Errorf("Unexpected transition to state %s with action 'none'", state)
+		case start:
+			// The selfCheckpoint must always be active to ensure that it can perform its functions in
+			// the face of a full control plane restart.
+			starts = append(starts, cs.selfCheckpoint.name)
+		case stop:
+			// If the checkpointer is stopped then stop all checkpoints. Next cycle they may restart.
+			for name, cp := range cs.checkpoints {
+				if cp.state.action() != none {
+					stops = append(stops, name)
 				}
+			}
+			stops = append(stops, cs.selfCheckpoint.name)
+			return starts, stops, removes
+		case remove:
+			// If the checkpointer is removed then remove all checkpoints, putting the selfCheckpoint
+			// last and removing all state.
+			for name, cp := range cs.checkpoints {
+				if cp.state.action() != none {
+					removes = append(removes, name)
+					delete(cs.checkpoints, name)
+				}
+			}
+			removes = append(removes, cs.selfCheckpoint.name)
+			delete(cs.checkpoints, cs.selfCheckpoint.name)
+			cs.selfCheckpoint = nil
+			return starts, stops, removes
+		default:
+			panic(fmt.Sprintf("unhandled action: %s", cs.selfCheckpoint.state.action()))
+		}
+	}
 
-				delete(activeCheckpoints, id)
+	// Update states for all the checkpoints and compute which to start / stop / remove.
+	for name, cp := range cs.checkpoints {
+		state := cp.state.transition(now, apiState{
+			apiAvailable: apiAvailable,
+			apiParent:    apiParentPods[name] != nil,
+			localRunning: localRunningPods[name] != nil,
+			localParent:  localParentPods[name] != nil,
+		})
+
+		if state != cp.state {
+			// Apply state transition.
+			// TODO(diegs): always apply this.
+			if cp.state.action() != state.action() {
+				switch state.action() {
+				case none:
+					glog.Errorf("Unexpected transition to state %s with action 'none'", state)
+				case start:
+					starts = append(starts, cp.name)
+				case stop:
+					stops = append(stops, cp.name)
+				case remove:
+					removes = append(removes, cp.name)
+					delete(cs.checkpoints, cp.name)
+				default:
+					panic(fmt.Sprintf("unhandled action: %s", state.action()))
+				}
 			}
+
+			glog.Infof("Checkpoint %s transitioning from state %s -> state %s", cp, cp.state, state)
+			cp.state = state
+		}
+	}
+
+	return starts, stops, removes
+}
+
+// stateSelfCheckpoint represents a checkpoint of the checkpointer itself, which has special
```

Comment:
maybe define all these states in a separate file? this `process.go` is getting big, I think it would be cleaner to just keep control logics here.

Reply:
Done